### ISSUE TITLE
Question about temporal mask in Wan

### ISSUE THREAD

**Issue by ZetangForward** [svg-project/Sparse-VideoGen#31]
Hi, I notice the same mask pattern in Wan, which is inconsistent with the image shown in the paper.
Specifically, the temporal mask pattern should be in a hardware efficient layout:

![Image](https://github.com/user-attachments/assets/c10ad3df-2bca-408e-a832-cf394135b7f6)

However, in the below code, the temporal mask looks the same as that of the spatial  mask:

https://github.com/svg-project/Sparse-VideoGen/blob/079364dc2e4ca6cd0c26c8c45eafeb9fbf51ef8e/svg/models/wan/utils.py#L87

I have tried with pseudo code (hunyuan code ... not Wan, but the core code and idea is same):

![Image](https://github.com/user-attachments/assets/7fb5a6cf-0e29-4308-911d-ff13c6f42cf5)

Can you explain this? 

@haochengxi @andy-yang-1

**Comment by haochengxi** [comment]
Hi, there's a reshape operation below, which acts as the layout transformation.

https://github.com/svg-project/Sparse-VideoGen/blob/079364dc2e4ca6cd0c26c8c45eafeb9fbf51ef8e/svg/models/wan/utils.py#L89

And this attention mask is used for simulation in our online profiling mechanism. So the attention mask is hardware inefficient.

**Comment by haochengxi** [comment]
If you are looking for the hardware-efficient calculation part, please refer to the code below.

These two lines are responsible for layout transformation:
https://github.com/svg-project/Sparse-VideoGen/blob/079364dc2e4ca6cd0c26c8c45eafeb9fbf51ef8e/svg/models/wan/attention.py#L227
https://github.com/svg-project/Sparse-VideoGen/blob/079364dc2e4ca6cd0c26c8c45eafeb9fbf51ef8e/svg/models/wan/attention.py#L231

We do the actual calculation here, where self.block_mask is hardware-efficient.
https://github.com/svg-project/Sparse-VideoGen/blob/079364dc2e4ca6cd0c26c8c45eafeb9fbf51ef8e/svg/models/wan/attention.py#L229

Hope this resolve your question :)

**Comment by haochengxi** [comment]
1. Yes, in the online profiling stage we need to explicitly construct an attention mask.
2. Yes, that's for layout transformation
3. FlexAttention can not directly use an attention mask as the input. You need to use the `create_block_mask` function to specify the sparsity pattern.
https://github.com/svg-project/Sparse-VideoGen/blob/079364dc2e4ca6cd0c26c8c45eafeb9fbf51ef8e/svg/models/wan/attention.py#L245

You can refer to this [blog ](https://pytorch.org/blog/flexattention/) from PyTorch team or my [Zhihu Blog](https://zhuanlan.zhihu.com/p/21829504838) for more information about flex-attention.

**Comment by ZetangForward** [comment]
> 1. Yes, in the online profiling stage we need to explicitly construct an attention mask.
> 2. Yes, that's for layout transformation
> 3. FlexAttention can not directly use an attention mask as the input. You need to use the `create_block_mask` function to specify the sparsity pattern.
>    
>      
>        
>          [Sparse-VideoGen/svg/models/wan/attention.py](https://github.com/svg-project/Sparse-VideoGen/blob/079364dc2e4ca6cd0c26c8c45eafeb9fbf51ef8e/svg/models/wan/attention.py#L245)
>        
>        
>             Line 245
>          in
>          [079364d](/svg-project/Sparse-VideoGen/commit/079364dc2e4ca6cd0c26c8c45eafeb9fbf51ef8e)
>        
>      
>      
>        
>    
>            
>              
>               block_mask = create_block_mask_cached(mask_mod, None, None, seq_len, seq_len, device=device, _compile=True)
> 
> You can refer to this [blog ](https://pytorch.org/blog/flexattention/) from PyTorch team or my [Zhihu Blog](https://zhuanlan.zhihu.com/p/21829504838) for more information about flex-attention.

Thanks, I will check the document source you provided first.
